### ISSUE TITLE
Batch Normalization

### ISSUE THREAD

**Issue by reunanen** [nyanp/tiny-cnn#147]
The benefits offered by [Batch Normalization](http://jmlr.org/proceedings/papers/v37/ioffe15.pdf) should be very helpful, especially when training on a CPU.

**Comment by nyanp** [comment]
@reunanen 
I agree, batch normalization is very important component in modern dnn architecture. Anyone who contribute it welcome.

**Comment by kingloo2014** [comment]
How to implete it in the tinycnn?

**Comment by nyanp** [comment]
@kingloo2014
I assume that we need a refactoring before implementing BN, because currently each layer take data one-by-one, not as mini-batch.

1. Refactor layer's API to handle mini-batch data at once
1. Implement forward/backward pass  (porting from caffe)
1. Implement save/load BN's params

**Comment by bhack** [comment]
A PR refactoring layer's API to handle mini-batch data at once would be a good step ahead.

**Comment by reunanen** [comment]
I can try to take a stab at this refactoring.

**Comment by reunanen** [comment]
The refactoring is now generally done and [merged](https://github.com/nyanp/tiny-cnn/pull/181).

Anyone willing to have a look at [items 2. and 3.](https://github.com/nyanp/tiny-cnn/issues/147#issuecomment-225482682), i.e. the actual BN implementation, please go ahead! (I know I won't have time anywhere in the near future.)

**Comment by edgarriba** [comment]
@reunanen Please, check short-term roadmap
https://github.com/nyanp/tiny-cnn/issues/186#issuecomment-227743683

**Comment by nyanp** [comment]
@reunanen @edgarriba @bhack 
I've added BN in 07b3146 and confirmed by a small experiment on MNIST that inserting BN layers actually improves training time. I've also tested that outputs of fprop/bprop has same values to caffe's BatchNormLayer.

![ebaf0eee744c3587e744e8bef27c1b50](https://cloud.githubusercontent.com/assets/378738/16899588/57f3e088-4c44-11e6-93bf-6d691fac799c.png)

Network models of BN version: (its based on example code in examples/main.cpp and just inserted bn layer for each conv layers)
```cpp
    nn << convolutional_layer<tan_h>(32, 32, 5, 1, 6) // 32x32 in, 5x5 kernel, 1-6 fmaps conv
       << batch_normalization_layer(28*28, 6)
       << average_pooling_layer<tan_h>(28, 28, 6, 2) // 28x28 in, 6 fmaps, 2x2 subsampling
       << convolutional_layer<tan_h>(14, 14, 5, 6, 16,
                                     connection_table(connection, 6, 16)) // with connection-table
       << batch_normalization_layer(10*10, 16)
       << average_pooling_layer<tan_h>(10, 10, 16, 2)
       << convolutional_layer<tan_h>(5, 5, 5, 16, 120)
       << fully_connected_layer<tan_h>(120, 10);
```

**Comment by reunanen** [comment]
@nyanp - Great!
